[sgen] Dynamic nursery sizing

BrzVlad opened the pull request:
Aim to use a nursery that is as big as possible, while avoiding too much memory growth and larger nursery pauses. Bigger nurseries increase performance by doing less repetitive minor work (stw, pinning, root scanning etc), reducing promotion ratio and overall being more parallelisation friendly.

We allocate a big nursery from the start, relying on the fact that virtual space allocation is cheap and it's not committed until the heap also grows to a significant size (when we can also increase the nursery). We should consider releasing physical memory when  the nursery shrinks, but this adds several complications with very little potential gain.

kumpera commented on mono/sgen/sgen-gc.c:
```diff
@@ -1346,6 +1346,16 @@ scan_copy_context_for_scan_job (void *worker_data_untyped, ScanJob *job)
 {
 	WorkerData *worker_data = (WorkerData *)worker_data_untyped;
 
+	if (!job->ops) {
```
What does this change tried to achieve exactly?

BrzVlad replied:
@kumpera This commit was added when I experimented with changing the context on the fly in order to speed up the finishing of the collection, at the cost of leaving unfinished business. I included it though since the worker ops are not fixed. They can be changed to serial when there is only one worker left, or we go through a suspend/resume step (when interrupting the the concurrent mark due to excessive heap growth) after which we change the context to one that also does copying.

While this commit doesn't really have an impact in practice, it conceptually unifies jobs and idle gray stack draining to "random work that needs to be executed with the worker ops, whatever those are at a given point"

kumpera replied:
Sounds good.

kumpera commented on mono/sgen/sgen-gc.c:
```diff
@@ -2999,13 +3023,187 @@ parse_double_in_interval (const char *env_var, const char *opt_name, const char
+ * potentially overriding them.
+ */
+static void
+init_sgen_mode (SgenMode mode, SgenMinor minor, SgenMajor major)
```
Hey Vlad,

There are a few issues with this commit:

1) Lots of magic numbers (plus on previous commit). All those numbers must go to sgen-conf.h as defines. Things like the default mode, default value for sgen_max_pause_time and sgen_max_pause_margin.

2) Using a GC mode should be exclusionary with manual tweaking. Letting them be mixed makes little sense.

3) I don't think we should add all those modes, specially for those we don't have a clear answer. Power and Memory are examples of superfluous modes we can avoid for now.

4) Some modes, like pause, should be parametric, the use should say what's the goal. 30ms pause time is an arbitrary number you picked that doesn't necessarily reflect what the user can expect to meet.

5) This is missing an update to the man page.

6) I don't agree on the configuration picks.

We can ignore HAVE_CONC_GC_AS_DEFAULT when using gc modes as there's no legacy behind it.

Balanced, I agree on this policy. Its goal is to keep pause times in check at a minimal cost in throughput.

Throughput, this one looks wrong, there's no point in using a concurrent GC when the goal is shorter wall clock. Parallel minor and major should be used. It should be give much larger nursery allowance so it better scales with major size.

Pause, I agree on this policy. Its goal is to try to meet a specific pause time goal at a modest sacrifice in performance.

As we discussed, we'll later have a more aggressive variant of the Pause goal where a bigger performance sacrifice is to be expected.

BrzVlad commented:
@kumpera I've removed the gc modes for now since we'll have more discussion on that topic later. I'll add them with a separate PR once we merge this. Any other comments on the dynamic nursery sizing alone (I've extracted a few variables to sgen-conf defines)

kumpera commented:
The dynamic nursery part LGTM
